WalletWasabi/Bases/PeriodicRunner.cs

```diff
@@ -1,132 +1,97 @@
-		private void ResetLastException()
-		{
-			LastException = null;
-			LastExceptionCount = 0;
-			LastExceptionFirstAppeared = DateTimeOffset.MinValue;
-		}
+		private volatile TaskCompletionSource<bool>? _tcs;
 
+		private LastExceptionTracker ExceptionTracker { get; }
+
+		/// <summary>
+		/// Normally, <see cref="ActionAsync(CancellationToken)"/> user-action is called every time that <see cref="Period"/> elapses.
+		/// This method allows to expedite the process by interrupting the waiting process.
+		/// </summary>
+		/// <remarks>
+		/// If <see cref="ExecuteAsync(CancellationToken)"/> is not actually in waiting phase, this method call makes
+		/// sure that next waiting process will be omitted altogether.
+		/// </remarks>
 		public void TriggerRound()
 		{
-			lock (TriggerLock)
-			{
-				TriggeringCts?.Cancel();
-			}
+			// Note: All members of TaskCompletionSource<TResult> are thread-safe and may be used from multiple threads concurrently.
+			_tcs?.TrySetResult(true);
 		}
 
+		/// <summary>
+		/// Abstract method that is called every <see cref="Period"/> or sooner when <see cref="TriggerRound"/> is called.
+		/// </summary>
 		protected abstract Task ActionAsync(CancellationToken cancel);
 
 		/// <inheritdoc />
 		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
 		{
 			while (!stoppingToken.IsCancellationRequested)
 			{
+				_tcs = new TaskCompletionSource<bool>();
+
 				try
 				{
+					// Do user action.
 					await ActionAsync(stoppingToken).ConfigureAwait(false);
-					LogAndResetLastExceptionIfNotNull();
+
+					// Log previous exception if any.
+					LogLastException(ExceptionTracker.LastException);
 				}
 				catch (Exception ex) when (ex is OperationCanceledException || ex is TaskCanceledException || ex is TimeoutException)
 				{
 					Logger.LogTrace(ex);
 				}
 				catch (Exception ex)
 				{
-					// Only log one type of exception once.
-					if (LastException is null // If the exception never came.
-						|| ex.GetType() != LastException.GetType() // Or the exception have different type from previous exception.
-						|| ex.Message != LastException.Message) // Or the exception have different message from previous exception.
-					{
-						// Then log and reset the last exception if another one came before.
-						LogAndResetLastExceptionIfNotNull();
-						// Set new exception and log it.
-						LastException = ex;
-						LastExceptionFirstAppeared = DateTimeOffset.UtcNow;
-						LastExceptionCount = 1;
-						Logger.LogError(ex);
-					}
-					else
-					{
-						// Increment the exception counter.
-						LastExceptionCount++;
-					}
+					// Exception encountered. Process it and log it, if it is first in the row.
+					LogLastException(ExceptionTracker.Process(ex));
+
+					Logger.LogError(ExceptionTracker.LastException!.Exception);
+				}
+
+				// Wait for the next round.
+				try
+				{
+					await Task.WhenAny(_tcs.Task, Task.Delay(Period, stoppingToken)).ConfigureAwait(false);
```

Comment:
I am not sure what happens with all the abandoned `_tcs.Task`. I assume once the GC removes the _tcs its task's Dispose is invoked, otherwise we would end with tons of tasks doing nothing and never completed.
So, if you understand exactly what happen here, ok; otherwise cancel the _tcs.Task.

Reply:
> So, if you understand exactly what happen here, ok; otherwise cancel the _tcs.Task.

* https://devblogs.microsoft.com/pfxteam/do-i-need-to-dispose-of-tasks/
* https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.dispose?view=netcore-3.1#System_Threading_Tasks_Task_Dispose

TLDR: It should not be necessary. Unless it's very performance sensitive applications, PeriodicRunner does not fall into that category.

Reply:
The first article says:
> This made the answer to the “should I dispose” question slightly difficult, as if you had a lot of tasks being used with WaitAll/WaitAny, it might actually have been good to dispose of those tasks.

The problem is that that article is pretty old and there were no `Task.WhenAll/Any` at that time so, it probably applies to these too. My concern is because the taks of the non-referenced `tcs` never complete and I have no idea what happen with them then.

Reply:
https://source.dot.net/#System.Private.CoreLib/Task.cs,1586 - This is what you have in mind, I think. Right, I'll try to tackle it.